src/main/java/org/spongepowered/common/world/FakePlayer.java

```diff
@@ -0,0 +1,203 @@
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.data.type.HandTypes;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.block.InteractBlockEvent;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.cause.NamedCause;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.util.Direction;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.SpongeImpl;
+import org.spongepowered.common.event.tracking.PhaseContext;
+import org.spongepowered.common.event.tracking.phase.plugin.PluginPhase;
+import org.spongepowered.common.interfaces.world.IMixinWorldServer;
+import org.spongepowered.common.registry.provider.DirectionFacingProvider;
+import org.spongepowered.common.util.VecHelper;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A marker to state that this is a fake player.
+ */
+public interface FakePlayer {
+
+    Controller controller = new Controller(SpongeImpl.getGame().createFakePlayerFactory());
+
+    public class Factory {
+
+        private final Map<UUID, EntityPlayerMP> players = new MapMaker().weakValues().makeMap();
+
+        public EntityPlayerMP getPlayer(WorldServer world, GameProfile profile) {
+            EntityPlayerMP player = players.get(profile.getId());
+            if (player == null) {
+                players.put(profile.getId(), player = new SimulatedPlayer(world, profile));
+            }
+            return player;
+        }
+
+    }
+
+    public static class Controller {
+
+        private final Factory factory;
+
+        Controller(Factory factory) {
+            this.factory = factory;
+        }
+
+        private EntityPlayerMP setUpPlayer(GameProfile profile, World world, int x, int y, int z, ItemStack itemInHand, Cause cause) {
+            EntityPlayerMP player = factory.getPlayer((WorldServer) world, profile);
+            player.world = (WorldServer) world;
+            player.interactionManager.setWorld((WorldServer) world);
+            player.posX = x;
+            player.posY = y;
+            player.posZ = z;
+            player.onGround = true;
+            net.minecraft.item.ItemStack stack = (net.minecraft.item.ItemStack) itemInHand;
+            if (stack == null) {
+                stack = net.minecraft.item.ItemStack.EMPTY;
+            }
+            player.inventory.setItemStack(stack);
+            player.inventory.mainInventory.set(player.inventory.currentItem, stack);
+
+            if (!cause.containsNamed(NamedCause.SOURCE)) {
+                // Cause tracker REQUIRES source?
+                cause = cause.with(NamedCause.of(NamedCause.SOURCE, cause.get(NamedCause.PLAYER_SIMULATED, Object.class).get()));
+            }
+            // Set up cause capture
+            ((IMixinWorldServer) world).getCauseTracker().switchToPhase(PluginPhase.State.FAKEPLAYER, PhaseContext.start()
```

Comment:
I originally wrote this before we had a cause tracker, once we did, I adapted it to use a [cause stack](https://github.com/SpongePowered/SpongeCommon/commit/361a6017#diff-feadb8e465f1fe25208169de8ca12a02R102). After the cause tracking refactor I changed it to use a phase (I should mention that Idon't really know how the current system works).

Essentially, the following needs to happen:
```
> API call (e.g. digBlock)
>> Setup fake player
>> Push given cause onto stack
>>> Make internal call
    * All events fired during this call will have the cause on the top of the stack *
    * another call to digBlock should be possible here because we use a stack *
<< Pop cause stack
<< Tear down fake player
< Return result
```

Maybe someone more knowledgeable on the cause tracking system will be able to see if I have done this correctly